paasta_tools/long_running_service_tools.py

```diff
@@ -345,20 +360,47 @@ def limit_instance_count(self, instances: int) -> int:
         return max(self.get_min_instances(), min(self.get_max_instances(), instances))
 
     def get_autoscaling_params(self) -> AutoscalingParamsDict:
-        default_params: AutoscalingParamsDict = {
-            "metrics_provider": "cpu",
+        default_provider_params: MetricsProviderDict = {
+            "type": MetricsProviderType.CPU,
             "decision_policy": "proportional",
             "setpoint": DEFAULT_AUTOSCALING_SETPOINT,
         }
-        return deep_merge_dictionaries(
-            overrides=self.config_dict.get("autoscaling", AutoscalingParamsDict({})),
-            defaults=default_params,
+
+        params = copy.deepcopy(
+            self.config_dict.get("autoscaling", AutoscalingParamsDict({}))
         )
+        if "metrics_providers" not in params or len(params["metrics_providers"]) == 0:
+            params["metrics_providers"] = [default_provider_params]
+        else:
+            for i in range(0, len(params["metrics_providers"])):
+                params["metrics_providers"][i] = deep_merge_dictionaries(
+                    overrides=params["metrics_providers"][i],
+                    defaults=default_provider_params,
+                )
+        return params
 
     def get_autoscaling_max_instances_alert_threshold(self) -> float:
         autoscaling_params = self.get_autoscaling_params()
         return autoscaling_params.get(
-            "max_instances_alert_threshold", autoscaling_params["setpoint"]
+            # TODO this default doesn't make sense for metrics providers that don't use setpoint
```

Comment:
@EvanKrall what are your thoughts here?

i can see us resolving this in two ways:

1.  look at the autoscaling params and use the `desired_active_requests_per_replica` if present
2.  give up on `desired_active_requests_per_replica` and rename it to `setpoint` so that the code is neater" (at the cost of people sometimes getting confused as to what setpoint means :p)

Reply:
It is kind of odd that for active-requests we have some positive number 0-infinity, where for other providers `setpoint` is typically a number between 0-1 expressed as a ratio of some maximum resource utilization that's possible (100% worker util, 100% cpu, etc). We _could_ maybe change active-requests to have two parameters: `max_active_requests_per_replica` (which would be some larger number than `desired` is currently) and `setpoint`. But that's probably just adding complexity where it isn't necessary.

With my [branch](https://github.com/Yelp/paasta/pull/3850) we could have setpoint be a large number for active requests and 0-1 for other providers, and then max_instances_alert_threshold for each of those would also follow the same rules.